openai_agents/memory/postgres_session.py

```diff
@@ -0,0 +1,297 @@
+            request.session_id,
+        )
+        for item in request.items:
+            # Use recursive conversion to handle nested objects
+            item_dict = _convert_to_json_serializable(item)
+
+            await conn.execute(
+                f"INSERT INTO {request.config.messages_table} (session_id, message_data) VALUES ($1, $2)",
+                request.session_id,
+                json.dumps(item_dict),
+            )
+
+    idempotence_helper = IdempotenceHelper(table_name=request.config.idempotence_table)
+    await idempotence_helper.idempotent_update(conn, add_items)
+
+
+class PostgresSessionPopItemRequest(BaseModel):
+    config: PostgresSessionConfig
+    session_id: str
+
+
+class PostgresSessionPopItemResponse(BaseModel):
+    item: TResponseInputItem | None
+
+
+@activity.defn
+async def postgres_session_pop_item_activity(
+    request: PostgresSessionPopItemRequest,
+) -> PostgresSessionPopItemResponse:
+    """Pop item from the session."""
+    conn = PostgresSession._get_connection()
+
+    async def pop_item(conn: asyncpg.Connection):
+        row = await conn.fetchrow(
+            f"WITH updated AS (UPDATE {request.config.messages_table} SET deleted_at = NOW() WHERE session_id = $1 AND operation_id = (SELECT operation_id FROM {request.config.messages_table} WHERE session_id = $1 AND deleted_at IS NULL ORDER BY operation_id DESC LIMIT 1) RETURNING message_data) SELECT message_data FROM updated",
+            request.session_id,
+        )
+        if row:
+            return PostgresSessionPopItemResponse(item=json.loads(row["message_data"]))
+        else:
+            return PostgresSessionPopItemResponse(item=None)
+
+    idempotence_helper = IdempotenceHelper(table_name=request.config.idempotence_table)
+    return await idempotence_helper.idempotent_update(conn, pop_item)
+
+
+class PostgresSessionClearSessionRequest(BaseModel):
+    config: PostgresSessionConfig
+    session_id: str
+
+
+@activity.defn
+async def postgres_session_clear_session_activity(
+    request: PostgresSessionClearSessionRequest,
+) -> None:
+    """Clear all items for this session."""
+    conn = PostgresSession._get_connection()
+
+    async def clear_session(conn: asyncpg.Connection):
+        await conn.execute(
+            f"UPDATE {request.config.messages_table} SET deleted_at = NOW() WHERE session_id = $1 AND deleted_at IS NULL",
+            request.session_id,
+        )
+
+    idempotence_helper = IdempotenceHelper(table_name=request.config.idempotence_table)
+    await idempotence_helper.idempotent_update(conn, clear_session)
+
+
+class PostgresSession(SessionABC):
+    """PostgreSQL-based implementation of session storage using operation_id ordering."""
+
+    def __init__(
+        self,
+        session_id: str,
+        config: PostgresSessionConfig,
+    ):
+        self.session_id = session_id
+        self.config = config
+
+    async def get_items(self, limit: int | None = None) -> list[TResponseInputItem]:
```

Comment:
If all of these items will end up in history anyways as a result of this call, is there any value in using a separate database instead of in-memory?